feat: Add GitLab CI provenance

wlynch opened the pull request:
<!-- What / Why -->
<!-- Describe the request in detail. What it does and why it's being changed. -->

This is a first pass at provenance generation for GitLab CI.

**Note:** I'm a noob when it comes to javascript, so feel free to call out any beginner mistakes / common practices 😅 

This is based loosely off of existing GitLab provenance documents, with some differences:

https://about.gitlab.com/blog/2022/11/30/achieve-slsa-level-2-compliance-with-gitlab/ https://gist.github.com/wlynch/c7fd8f53adc77d3c0ec82356e4d43cb5

Currently this pulls values from [environment variables](https://docs.gitlab.com/ee/ci/variables/predefined_variables.html). I'm aware we want to pull most data from authenticated JWTs for GitHub provenance, but I don't know what is in-flight so I am starting here for now, and marking as draft to get early feedback. @bdehamer happy to sync up and coordinate on changes you're making on the GitHub side.

We're doing things slightly out of order while we're waiting for things to land on the [Sigstore side to recognize GitLab issued JWTs](https://github.com/sigstore/fulcio/issues/243), so this has not been tested e2e in a real environment. Marking this as v1alpha1 until we have more confidence in the provenance spec + generation. If there's a better way to denote potential instability / experimental behavior let me know.
Worst case this should just error out when trying to fetch a Sigstore cert.

## References
<!-- Examples:
  Related to #0
  Depends on #0
  Blocked by #0
  Fixes #0
  Closes #0
-->

Fixes #6373

wlynch committed:
libnpmpublish: Add GitLab CI provenance.

This is a first pass at provenance generation for GitLab CI.

This is based loosely off of existing GitLab provenance documents:
https://about.gitlab.com/blog/2022/11/30/achieve-slsa-level-2-compliance-with-gitlab/
https://gist.github.com/wlynch/c7fd8f53adc77d3c0ec82356e4d43cb5

Currently this pulls values from environment variables.
I'm aware we want to pull this data from authenticated JWTs for
GitHub provenance, but I don't know what is in flight so I am starting
here for now, marking as v1alpha1 until we have more confidence in the
provenance spec.

ljharb commented on workspaces/libnpmpublish/lib/provenance.js:
```diff
@@ -63,7 +82,142 @@ const generateProvenance = async (subject, opts) => {
   }
 
-  return sigstore.attest(Buffer.from(JSON.stringify(payload)), INTOTO_PAYLOAD_TYPE, opts)
+  return Buffer.from(JSON.stringify(payload))
```
the buffer and json parts seem to apply to both, should they be hoisted out to the caller too?

wlynch replied:
Done.

wraithgar commented on workspaces/libnpmpublish/lib/provenance.js:
```diff
@@ -1,15 +1,34 @@
+  var provenance
+  switch (true) {
+    case ci.GITHUB_ACTIONS:
+      provenance = githubProvenance(subject)
```
These are the kinds of things I was talking about avoiding when I suggested an `if` statement.  There is no need for single-use functions quite yet.  Just an if statement with everything that used to be here in one case, and the gitlab stuff in the next one

wlynch replied:
Done.

wraithgar commented on workspaces/libnpmpublish/lib/publish.js:
```diff
@@ -144,21 +144,24 @@ const buildMetadata = async (registry, manifest, tarballData, spec, opts) => {
-        new Error('Provenance generation in GitHub Actions requires "write" access to the "id-token" permission'),
-        { code: 'EUSAGE' }
-      )
+    switch (true) {
```
If statements here are likely to be much easier to follow.

wlynch replied:
Done.

wlynch committed:
libnpmpublish: [gitlab] un-refactor generate provenance, add check for
SIGSTORE_ID_TOKEN.

wlynch reviewed (commented):
Thanks for the feedback! Updated.

Will move out of draft once I hear more about plans for using JWTs in provenance.

wraithgar commented on workspaces/libnpmpublish/lib/provenance.js:
```diff
@@ -1,68 +1,208 @@
+        ],
+      },
+    }
+  } else {
```
lib/publish.js already tests this so this is a pathway we can't even get to from this module's api right?

wlynch replied:
Assuming it's always called via publish.js, yes. I included this for completeness / defense in depth, but if you feel it isn't necessary lmk and I'll remove it.

wraithgar replied:
Historically code like this has lead to a lot of dead code in npm specifically.  Our current testing approach is to test the api itself, you can remove this and the unit test.

wlynch replied:
Done.

wraithgar commented on workspaces/libnpmpublish/test/publish.js:
```diff
@@ -9,6 +9,7 @@ const t = require('tap')
 
 const MockRegistry = require('@npmcli/mock-registry')
 const mockGlobals = require('../../../test/fixtures/mock-globals.js')
+const { generateProvenance } = require('../lib/provenance.js')
```
We are avoiding unit tests here. If the code can't be reached from `require('libnpmpublish')` it's code that shouldn't exist.

wlynch replied:
Done.

feelepxyz commented on workspaces/libnpmpublish/lib/publish.js:
```diff
@@ -144,19 +144,27 @@ const buildMetadata = async (registry, manifest, tarballData, spec, opts) => {
+      if (!process.env.SIGSTORE_ID_TOKEN) {
+        throw Object.assign(
+          /* eslint-disable-next-line max-len */
+          new Error('Provenance generation in GitLab CI requires "SIGSTORE_ID_TOKEN" with "sigstore" audience to be present in "id_tokens"'),
```
As a user, how would I remedy this issue? Will it be enabled by default on latest version of gitlab?

wlynch replied:
This is a field users set in their `.gitlab-ci.yml` file. Added the link to the GitLab documentation!

If/when there's more specific Sigstore documentation we can point there instead, but tl;dr it needs to look like:

```yaml
job:
  id_tokens:
    SIGSTORE_ID_TOKEN:
      aud: sigstore
```

feelepxyz commented on workspaces/libnpmpublish/lib/provenance.js:
```diff
@@ -1,68 +1,208 @@
+const GITHUB_BUILD_TYPE_VERSION = 'v2'
+
+const GITLAB_BUILD_TYPE_PREFIX = 'https://github.com/npm/cli/gitlab'
+const GITLAB_BUILD_TYPE_VERSION = 'v1alpha1'
```
Would prefix `v0` so it doesn't clash with a future `v1`

```suggestion
const GITLAB_BUILD_TYPE_VERSION = 'v0alpha1'
```

wlynch replied:
Done.

feelepxyz commented on workspaces/libnpmpublish/lib/provenance.js:
```diff
@@ -1,68 +1,208 @@
+      predicateType: SLSA_PREDICATE_TYPE,
+      predicate: {
+        buildType: `${GITLAB_BUILD_TYPE_PREFIX}/${GITLAB_BUILD_TYPE_VERSION}`,
+        builder: { id: `${env.CI_PROJECT_URL}/-/runners/${env.CI_RUNNER_ID}` },
```
Looking at the [claims mapped here](https://github.com/sigstore/fulcio/pull/1097/files), do you have access to the `CI_RUNNER_ENVIRONMENT` here so we could use that instead of the id? Or is id the env? This would make it easier to write policies trust only the `gitlab-hosted` runner for example

feelepxyz replied:
Also should this be per project or just set to the server url runner to make it easier to have some kind of stable identifier for all gitlab.com hosted runners for example?

? `env.CI_SERVER_URL/-/runners/${env. CI_RUNNER_ENVIRONMENT}`

Brcrwilliams replied:
Looking at https://github.com/in-toto/attestation/blob/v0.1.0/spec/predicates/provenance.md#fields, I don't think that the runner id is an appropriate value to use here.

> The identity MUST reflect the trust base that consumers care about. How detailed to be is a judgement call. For example, [GitHub Actions](https://github.com/in-toto/attestation/blob/v0.1.0/spec/predicates/provenance.md#github-actions) supports both GitHub-hosted runners and self-hosted runners. The GitHub-hosted runner might be a single identity because, it's all GitHub from the consumer's perspective. Meanwhile, each self-hosted runner might have its own identity because not all runners are trusted by all consumers.

The runner id identifies the specific machine, and we cycle through machines quite often. For gitlab-hosted runners, we probably want this to be a generic value that denotes the runner is hosted by GitLab. For self-hosted runners, we probably want some way to map this back to the instance, group, or project that is hosting the runner.

wlynch replied:
I'm generally in favor, but I think there's some desire to have this match the existing [GitLab provenance document](https://about.gitlab.com/blog/2022/11/30/achieve-slsa-level-2-compliance-with-gitlab/), which uses CI_RUNNER_ID (see https://gist.github.com/wlynch/c7fd8f53adc77d3c0ec82356e4d43cb5 for an example provenance doc I pulled that I used as the basis for this). 

This is a common theme for why fields are the way they are for this provenance document.

I think my preference would to start with parity with the existing GitLab SLSA v0.2 spec, and I can sync with @Brcrwilliams @marshall007 and other GitLab folks to how we want this to change this moving forward.

ianlewis replied:
The [docs](https://slsa.dev/provenance/v0.2#builder) say `builder.id` defines the "trust base that consumers care about". I can't say too much about GitLab without more knowledge of that platform, but generally I think this should identify something like "GitLab hosted runner" or "GitLab custom runner". I don't think the runner ID will identify a significant difference in the "trust base".

/cc @MarkLodato @kpk47

However, wrt the format overall maybe this is best taken up with the GitLab [Supply Chain Security working group](https://about.gitlab.com/company/team/structure/working-groups/software-supply-chain-security/) of which I think @kpk47 is a member.

feelepxyz replied:
Keeping parity with the existing GitLab SLSA v0.2 spec makes sense to me. Sounds like it would also be useful to start an effort to figure out how the 1.0 spec should look for GitLab and get this included in the slsa org.

Before you can actually test this against the npm registry, we'll need to update the server side checks to accept gitlab provenance statements. There's currently a simple allow list of issuers to allow us to gradually roll out support, as well as checks to make sure the extensions in the signing cert matches what's in the provenance statement.

@wlynch to aid with registry integration, could you include a complete `.sigstore` bundle files for GitLab? It's a bit annoying getting the sigstore bundle at the moment, but you could modify this line and write the file to disc during a pipeline run: https://github.com/npm/cli/blob/a558bbd3760e8da272d7afdd11e7b0f1238f92dc/workspaces/libnpmpublish/lib/publish.js#L185

Ideally we would also have the fulcio certificate updated to include all the new gitlab claims but can start without this.

wlynch replied:
> However, wrt the format overall maybe this is best taken up with the GitLab Supply Chain Security working group

I already brought it up in the WG. 🙂 There's general interest in working towards improving the existing SLSA v0.2 spec + moving to v1.0. We can follow up there.

> @wlynch to aid with registry integration, could you include a complete .sigstore bundle files for GitLab?

I'll do this as soon as https://github.com/sigstore/fulcio/pull/983 goes live!

laurentsimon replied:
One additional problem of using `builder: { id: ${env.CI_PROJECT_URL}/-/runners/${env.CI_RUNNER_ID}}` as builder.id is that every project has a different builder. This kinda defeats the purpose of SLSA which scales by having a small set of builders, which you can then use for policy enforcement during verification. The builder.id, as written in this PR, would need to be generated dynamically for verification, even though the builders are technically the same (for GitLab hosted, I mean). See a longer discussion in https://github.com/slsa-framework/slsa/issues/655.

feelepxyz commented on workspaces/libnpmpublish/lib/provenance.js:
```diff
@@ -1,68 +1,208 @@
+            digest: {
+              sha1: env.CI_COMMIT_SHA,
+            },
+            entryPoint: env.CI_JOB_NAME,
```
What is the job name set to? Is this unique for all pipeline configs in a project? Wondering if this should instead be set to a project relative path to the pipeline ref?

Brcrwilliams replied:
The job name is a user-defined name for the job. It is unique for all jobs in the pipeline.

I might be able to suggest what to use for this, but I'm not clear on the purpose of this field. I'm looking at https://github.com/in-toto/attestation/blob/v0.1.0/spec/predicates/provenance.md#fields and I see no mention of `.invocation.entryPoint`.

laurentsimon replied:
entryPoint defines the trigger script run, ie the very first user-defined script that is run to kick-off the build process

laurentsimon replied:
defined in v0.2 https://slsa.dev/provenance/v0.2

wlynch replied:
configsource / entrypoint was a slsa 0.2 field idk where the docs went in the intoto repo, but https://slsa.dev/provenance/v0.2#fields has some documentation on it.

feelepxyz commented on workspaces/libnpmpublish/lib/provenance.js:
```diff
@@ -1,68 +1,208 @@
-          digest: {
-            sha1: env.GITHUB_SHA,
+        metadata: {
+          buildInvocationId: `${env.CI_JOB_URL}`,
```
Will this be the same as? `server_url + project_path + /-/pipelines/ + pipeline_id` in the [oidc mappings](https://github.com/sigstore/fulcio/pull/1097/files)

Brcrwilliams replied:
No. A pipeline contains many jobs, and a job is related to a single pipeline. A job URL is at `server_url + project_path + /-/jobs/ + job_id` and from that page it's possible to see which pipeline it belongs to.

feelepxyz commented on workspaces/libnpmpublish/lib/provenance.js:
```diff
@@ -1,68 +1,208 @@
+            },
+            entryPoint: env.CI_JOB_NAME,
+          },
+          parameters: {
```
According to the [slsa 0.2 spec](https://slsa.dev/provenance/v0.2), `parameters` are external inputs that influenced the build on top of `invocation.configSource`.

Wondering if most of these should be moved to `environment` instead, as it's not clear how exactly they might influence the build.

Also, there's a lot here! Do we really need to include all these values? Could we prune this down to things that are crucial for provenance, i.e. understanding where and how it was produced.

Could you provide a complete example that might make it easier to spot what's useful?

cc @ianlewis @MarkLodato @laurentsimon would love your input here as well. Don't know how much time you've spent looking at how a provenance statement should look for gitlab? Maybe we should draft up a 1.0 spec for gitlab?

laurentsimon replied:
I'll +1 using v1.0 specs right away.

re: `parameters` on GitHub I think these would be the workflow inputs provided by the user, e.g., on a `workflow_dispatch` event. In our GitHub generators, we actually did not populate the field as it is already recorded via the environment's `event_payload` which contains  an `inputs` parameter. Maybe an oversight on our side.

laurentsimon replied:
> Also, there's a lot here!

Which of these will be forgeable? I think we can only include the fields that are part of the OIDC token and present in the Fulcio cert? Otherwise the npm registry won't be able to verify these values during publication

If there is something akin to a re-usable workflow on GitLab, we could record more.

wlynch replied:
This is another case of aiming for parity with existing GitLab SLSA v0.2 spec.

cc @marshall007 @Brcrwilliams for SLSA v1.0 plans.

wlynch replied:
GitLab does some auto-scrubbing to prevent secrets set in environment vars from leaking out.

These are explicitly listed as a simplified way to emulate that behavior so we're not accidentally including secret values.

ianlewis replied:
I agree these would generally be the "inputs" to a workflow. Everything `CI*` and `GITLAB_*` would be better placed in the `environment` imho.

This is much clearer in SLSA v1.0 where it's clearly delineated between parameters under user control (`externalParameters`) and those under the platform's control (`internalParameters`). In 0.2 it's a bit more fuzzy but `environment` generally maps to the `internalParameters` and `parameters` maps to `externalParameters`.

feelepxyz replied:
> GitLab SLSA v0.2 spec

Ah gotcha, keeping parity with the existing gitlab v0.2 schema makes sense to me unless we want to go to 1.0 but sounds like this might take some time to settle on 👍

feelepxyz commented:
@wlynch this is very exciting! Finally had some time to review this. I think we should also give the SLSA team some time to chime in here.

Part of me also wonders if we should just got to SLSA 1.0 for GitLab from the get go and write up a spec for this in the official repo [like was done for GHA](https://github.com/slsa-framework/github-actions-buildtypes/tree/main/workflow/v1). This would simplify our lives a bit on the npm registry as we would only need support 3 different predicate types for GitLab when rendering the UI (1.0) + GitHub (SLSA 0.2 + 1.0). 

Going to figure out if we can add support for SLSA 1.0 in the next few weeks, maybe up to a month.

wlynch committed:
GitLab provenance: change version to v0alpha1, include links to GitLab
OIDC docs.

wraithgar commented:
I updated this PR to be a merge request against the `provenance` branch. This is what was used during beta testing of the GitHub provenance generation and I am hoping it can prove useful to GitLab's testing.

Once this PR lands folks can `npm install -g npm/cli#provenance` and get the code as y'all are helping to refine it.  This is a way folks can test e2e in a real environment.

This will also lower the barrier to what is "acceptable" to land. We don't have to land the 100% finished product all at once, and we can have a little more wiggle room in case we need to make "breaking" changes in the branch before then.

wlynch commented:
SGTM! Happy to move this out of draft then. 🙂

wlynch committed:
Remove fallthrough case in provenance.js.

This case is already covered in publish.js, so it was recommended that
this was removed.

wraithgar commented:
Ok sorry for the rebase noise I think this PR looks right as far as destination branch.  We'll keep `provenance` synced w/ latest as best we can.  Folks using this branch to test GitLab provenance are opting in to testing something that's definitely in an alpha phase for now.

wraithgar commented on workspaces/libnpmpublish/package.json:
```diff
@@ -44,7 +44,7 @@
     "npm-registry-fetch": "^14.0.3",
     "proc-log": "^3.0.0",
     "semver": "^7.3.7",
-    "sigstore": "^1.0.0",
```
npm cli team needs to make this change in `latest`

wlynch replied:
Looks like there was some work towards this in https://github.com/npm/cli/commit/fae5e0063215e97ee18a60f8136a06045d621ec2.

Anything I can do to help this along? 👀

wraithgar replied:
It already landed in `latest` at that point, I think if you rebase your PR against `latest` it should clear.

wlynch replied:
I don't know if https://github.com/npm/cli/commit/fae5e0063215e97ee18a60f8136a06045d621ec2 covers the libnpmpublish workspace? package.json still shows 1.0.0 at latest: https://github.com/npm/cli/blob/0783cff9653928359a6c68c8fdf30b9fd02130c9/workspaces/libnpmpublish/package.json#L47

wraithgar replied:
Oh right thanks, got my wires crossed there

ianlewis commented on workspaces/libnpmpublish/lib/provenance.js:
```diff
@@ -1,68 +1,204 @@
+            RUNNER_GENERATE_ARTIFACTS_METADATA: env.RUNNER_GENERATE_ARTIFACTS_METADATA,
+          },
+          environment: {
+            name: env.CI_RUNNER_DESCRIPTION,
+            architecture: env.CI_RUNNER_EXECUTABLE_ARCH,
+            job: {
+              id: env.CI_JOB_ID,
+            },
```
I think these could just be 1:1 with their env names. I think that makes it easier for downstream consumers familiar with the GitLab environment.

feelepxyz commented on workspaces/libnpmpublish/lib/provenance.js:
```diff
@@ -1,68 +1,204 @@
+        builder: { id: `${env.CI_PROJECT_URL}/-/runners/${env.CI_RUNNER_ID}` },
+        invocation: {
+          configSource: {
+            uri: `git+${env.CI_PROJECT_URL}@${env.CI_BUILD_REF}`,
```
@wlynch I'm trying to map these predicate values in `configSource` to the [Fulcio OID mappings](https://github.com/sigstore/fulcio/pull/1097/files) so we could look at validating the certificate extensions against what's in the predicate to make sure they haven't been forged/spoofed during build.

- Does `CI_PROJECT_URL` map to the `server_url + project_path` claims set as the `Source Repository URI`?
- Does `CI_BUILD_REF` map to the `ref` claim set as the `Source Repository Ref`?
- Does `CI_COMMIT_SHA` map to the `sha` claim set as the `Source Repository Digest`
- Does `CI_JOB_NAME ` map to `pipeline_ref` claim that's set as the `Build Config URI`? 
- Does `CI_JOB_URL` map to the `server_url + project_path + /-/pipelines/ + pipeline_id` claims set as the `Run Invocation URI`?

I realise changing some of these to map would mean the provenance statement is no longer in line with the GitLab v0.2 schema, but any thoughts on getting this changed to align with the Fulcio changes?

We could punt on this consistency verification for GitLab and just pull these values from the cert for UI display but means someone could still easily mess with the build env, spoof details and someone inspecting the predicate on Rekor would see the incorrect values unless the also checked the cert and compared.

wlynch replied:
> Does CI_PROJECT_URL map to the server_url + project_path claims set as the Source Repository URI?

✅  Yes.

> Does CI_BUILD_REF map to the ref claim set as the Source Repository Ref?

🟡  Mostly. In the cert this will be the ref path (e.g. the full `refs/heads/main` path).

> Does CI_COMMIT_SHA map to the sha claim set as the Source Repository Digest

✅ Yes

> Does CI_JOB_NAME  map to pipeline_ref claim that's set as the Build Config URI?

❌ No. This would be the name of the job in the pipeline, but jobs are only addressable by id, not name.

> Does CI_JOB_URL map to the server_url + project_path + /-/pipelines/ + pipeline_id claims set as the Run Invocation URI?

❌ No this would match the Build Config URI.

wlynch replied:
Added additional fields in environment to capture pipeline ref + pipeline id.

feelepxyz commented on workspaces/libnpmpublish/lib/provenance.js:
```diff
@@ -1,68 +1,204 @@
-          environment: false,
-          materials: false,
+    }
+  }
+  if (ci.GITLAB) {
```
```suggestion
  } else if (ci.GITLAB) {
```

wlynch replied:
We can't do this because coverage will fail because the else path is unreachable since we have a separate if/else in publish.js.

wlynch committed:
Set GITHUB_ACTIONS environment variable to false in tests.

Current theory is that this is causing the tests to fail, since the
library is picking up the real GitHub Actions runner when running in CI.

wlynch committed:
Add additional pipeline data into environment.

This data probably shouldn't be in environment, but this is where
similar data lives in the GitLab provenance spec today so it likely
makes the most sense to co-locate.

wraithgar commented:
`https://github.com/npm/cli/pull/6416` will update sigstore to '^1.4.0' in libnpmpublish's package.json.

wlynch commented:
Rebased on latest to pull in latest dep updates (incl sigstore 1.4.0 which let me revert the package.json change). We may need someone to pull in `latest` to the `provenance` if we want to clean the diff here.

wraithgar commented:
I'll do that after the release. I'm relatively sure `latest` is done moving but things are gonna be like this on release day.

wlynch commented:
Rebased on latest provenance branch!

wraithgar commented on workspaces/libnpmpublish/lib/provenance.js:
```diff
@@ -1,68 +1,209 @@
-          uri: `git+${env.GITHUB_SERVER_URL}/${env.GITHUB_REPOSITORY}@${env.GITHUB_REF}`,
-          digest: {
-            sha1: env.GITHUB_SHA,
+  var payload
```
```suggestion
  let payload
```

wlynch replied:
Done.

wraithgar commented on workspaces/libnpmpublish/lib/publish.js:
```diff
@@ -144,19 +144,27 @@ const buildMetadata = async (registry, manifest, tarballData, spec, opts) => {
+      if (!process.env.SIGSTORE_ID_TOKEN) {
+        throw Object.assign(
+          /* eslint-disable-next-line max-len */
+          new Error('Provenance generation in GitLab CI requires "SIGSTORE_ID_TOKEN" with "sigstore" audience to be present in "id_tokens". See https://docs.gitlab.com/ee/ci/secrets/id_token_authentication.html for more info.'),
```
```suggestion
          new Error('Provenance generation in GitLab CI requires "SIGSTORE_ID_TOKEN" with "sigstore" audience to be present in "id_tokens". For more info see:\nhttps://docs.gitlab.com/ee/ci/secrets/id_token_authentication.html'),
```

This helps folks w/ screen readers find, copy, and paste the url in isolation.

wlynch replied:
Done.

wraithgar commented:
Gave some tiny code tweak suggestions.  We'd review and clean up the `provenance` branch before landing it in `latest` so things like that could be handled there if this branch is already otherwise ready to go.  Your call.

wlynch committed:
libnpmpublish/gitlab: var -> let, revise error string.

wlynch commented:
SGTM! Made the tweaks anyway :)

wraithgar reviewed (approved):
From a cli code perspective this is all fine. I'll let @feelepxyz decide when the implementation is appropriate to land in `provenance`

feelepxyz commented on workspaces/libnpmpublish/lib/provenance.js:
```diff
@@ -1,68 +1,209 @@
+            CI_PIPELINE_ID: env.CI_PIPELINE_ID,
+            CI_PIPELINE_IID: env.CI_PIPELINE_IID,
+            CI_PIPELINE_SOURCE: env.CI_PIPELINE_SOURCE,
+            CI_PIPELINE_URL: env.CI_PIPELINE_URL,
```
@wlynch could we add `pipeline_ref` and `pipeline_sha` [that are now in the claims](https://gitlab.com/gitlab-org/gitlab/-/issues/404722)? Or are these in here but with different names?

feelepxyz replied:
Ah I see these are still in progress: https://gitlab.com/gitlab-org/gitlab/-/merge_requests/117923

wlynch replied:
pipeline_ref is populated in `environent.pipeline.ref` via `CI_CONFIG_PATH`.

`pipeline_sha` is WIP on the GitLab side and may not always be populated (i.e. if being fetched from remote sources) - I've raised with GitLab folks. The safer thing to key off of for provenance checking is pipeline id / job id. I don't think this is blocking for this PR.

Brcrwilliams replied:
@wlynch `pipeline_ref` is actually a URI that links to the pipeline configuration. The name is easy to confuse with a git ref. 

@marshall007 @aladh Maybe we should call it something else like `pipeline_config_uri` and `pipeline_config_sha`.

Brcrwilliams replied:
@wlynch `pipeline_ref` is actually a URI that links to the pipeline configuration. The name is easy to confuse with a git ref. 

@marshall007 @aladh Maybe we should call it something else like `pipeline_config_uri` and `pipeline_config_sha`.

wlynch replied:
That's what we want. 🙂 

What we're ultimately looking for is a hook to the Fulcio cert Build Config URI to match for verification.

No objections to renaming though.

feelepxyz reviewed (commented):
This is looking good 👍  would be good to get the final changes from the claims to add values for `pipeline_ref` and `pipeline_sha` before we merge this.

wlynch committed:
Remove CI_REPOSITORY_URL, CI_BUILD_REF.

CI_BUILD_REF doesn't actually exist, CI_REPOSITORY_URL contains an
access token.

wlynch committed:
Remove GITLAB_USER_NAME, GITLAB_USER_EMAIL.

These values are a bit too much PII, so remove for now and only include
GITLAB_USER_ID and GITLAB_USER_LOGIN.

wlynch commented:
Sample provenance from Sigstore staging: https://gist.github.com/wlynch/42e89527d51bc72a61279f0c7f3be1cd

GitLab support landed in Fulcio prod today, so you should be able to test this out in prod as well!

feelepxyz commented on workspaces/libnpmpublish/lib/provenance.js:
```diff
@@ -1,68 +1,206 @@
+            CI_BUILD_REF_NAME: env.CI_BUILD_REF_NAME,
+            CI_BUILD_REF_SLUG: env.CI_BUILD_REF_SLUG,
+            CI_BUILD_STAGE: env.CI_BUILD_STAGE,
+            CI_COMMIT_AUTHOR: env.CI_COMMIT_AUTHOR,
+            CI_COMMIT_BEFORE_SHA: env.CI_COMMIT_BEFORE_SHA,
+            CI_COMMIT_BRANCH: env.CI_COMMIT_BRANCH,
+            CI_COMMIT_DESCRIPTION: env.CI_COMMIT_DESCRIPTION,
+            CI_COMMIT_MESSAGE: env.CI_COMMIT_MESSAGE,
```
On PII, looks like some of these include full name and email, do you want to remove these as well?

```
"CI_COMMIT_AUTHOR": "Full Name <email@chainguard.dev>",
"CI_COMMIT_DESCRIPTION": "\nSigned-off-by: Full Name <email@chainguard.dev>",
"CI_COMMIT_MESSAGE": "hack sigstore-js\n\nSigned-off-by: Full Name <email@chainguard.dev>\n",
```

wlynch replied:
These are from the Git commit, so we wouldn't *really* be able to hide these, but done!

wlynch committed:
Remove CI_COMMIT_AUTHOR, CI_COMMIT_DESCRIPTION, CI_COMMIT_MESSAGE

feelepxyz commented on workspaces/libnpmpublish/test/publish.js:
```diff
@@ -903,3 +903,244 @@ t.test('automatic provenance with incorrect permissions', async t => {
+
+t.test('publish existing package with provenance in gitlab', async t => {
+  // Environment variables
+  /*
+  const oidcURL = 'https://mock.oidc'
+  const requestToken = 'decafbad'
+  */
```
```suggestion
```

wlynch replied:
Done.

feelepxyz commented on workspaces/libnpmpublish/test/publish.js:
```diff
@@ -903,3 +903,244 @@ t.test('automatic provenance with incorrect permissions', async t => {
+    },
+  }
+
+  /*
+  const oidcSrv = MockRegistry.tnock(t, oidcURL)
+  oidcSrv.get('/?audience=sigstore', undefined, {
+    authorization: `Bearer ${requestToken}`,
+  }).reply(200, { value: idToken })
+  */
+
```
```suggestion
```

wlynch replied:
Done.

wlynch committed:
Remove commented lines from test.

feelepxyz commented on workspaces/libnpmpublish/test/publish.js:
```diff
@@ -903,3 +903,244 @@ t.test('automatic provenance with incorrect permissions', async t => {
+  t.ok(ret, 'publish succeeded')
+  t.match(log, [
+    ['notice', 'publish',
+      'Signed provenance statement with source and build information from GitHub Actions'],
```
Oh I think we need to update the log were we publish to handle gitlab as well:
```suggestion
      'Signed provenance statement with source and build information from GitLab CI'],
```

feelepxyz replied:
https://github.com/npm/cli/blob/baae2d60109c3362c38ad09c231a32c84c26f63a/workspaces/libnpmpublish/lib/publish.js#L207

wlynch replied:
Done.

wlynch committed:
Use ci-info.name for provenance log statement.

wlynch commented:
prod bundle: https://gist.github.com/wlynch/26bb5d8da80419b915b23f9c3a75bebe
pulled from https://gitlab.com/wlynch/npm-provenance-example/-/jobs/4276586778

feelepxyz reviewed (approved):
LGTM! I'm 👍  on merging this into the `provenance` branch.

@wraithgar would you like to do the honours?

wraithgar commented:
Yay!  Thanks @wlynch for seeing this through.  Thanks @feelepxyz for reviewing it.

Installing this version of npm should be as simple as: `npm install -g npm/cli#provenance`

wraithgar commented:
@wlynch Can we get a PR started to the docs to update `https://docs.npmjs.com/generating-provenance-statements` to include what the gitlab instructions should be?  We can wait to land it till we also land the provenance branch but let's not forget the docs.

https://github.com/npm/documentation/blob/e52916f65fc640287f269d0268b8f77c5e41b33d/content/packages-and-modules/securing-your-code/generating-provenance-statements.mdx

SiaraMist commented:
👋 If it would be helpful for a tech writer to work on the docs for this, I wrote the existing npm build provenance docs and am happy to add GitLab content too if there is detailed source material. If you would like me to add the updates, just let me know here or in Slack and I can get that process started. (No worries if you just want to make the updates yourself though.)
